lib/pbench/server/database/models/template.py

```diff
@@ -144,26 +167,51 @@ def __str__(self):
 
     def add(self):
         """
-        add Add the Template object to the database
+        Add the Template object to the database
         """
         try:
             Database.db_session.add(self)
             Database.db_session.commit()
-        except IntegrityError:
-            Template.logger.exception("Duplicate template {}", self.name)
-            raise TemplateDuplicate(self.name)
+        except IntegrityError as e:
+            # Postgres engine returns (code, message) but sqlite3 engine only
+            # returns (message); so always take the last element.
+            cause = e.orig.args[-1]
+            Database.db_session.rollback()
+            if cause.find("UNIQUE constraint") != -1:
+                Template.logger.warning("Duplicate template {!r}: {}", self.name, cause)
+                raise TemplateDuplicate(self.name)
+            elif cause.find("NOT NULL constraint") != -1:
+                Template.logger.warning("Missing parameter {!r}, {}", self.name, cause)
+                raise TemplateMissingParameter(self.name, cause)
+            else:
+                Template.logger.exception("Other integrity error {}", cause)
+                raise
         except Exception:
             self.logger.exception("Can't add {} to DB", str(self))
             Database.db_session.rollback()
             raise TemplateSqlError("adding", self.name)
 
     def update(self):
         """
-        update Update the database row with the modified version of the
+        Update the database row with the modified version of the
         Template object.
         """
         try:
             Database.db_session.commit()
+        except IntegrityError as e:
+            # Postgres engine returns (code, message) but sqlite3 engine only
+            # returns (message); so always take the last element.
+            cause = e.orig.args[-1]
+            Database.db_session.rollback()
+            if cause.find("UNIQUE constraint") != -1:
+                Template.logger.warning("Duplicate template {!r}: {}", self.name, cause)
+                raise TemplateDuplicate(self.name)
+            elif cause.find("NOT NULL constraint") != -1:
+                Template.logger.warning("Missing parameter {!r}, {}", self.name, cause)
+                raise TemplateMissingParameter(self.name, cause)
+            else:
+                Template.logger.exception("Other integrity error {}", cause)
+                raise
```

Comment:
I'm reluctant mostly because that adds another level of traceback to obscure the origin. At least, unless I simply return an exception to raise here, which defeats the clarity of the [re-]`raise` at the end. I decided I didn't really like it either way, although I'm also not enthusiastic about duplicate code.

Reply:
> that adds another level of traceback to obscure the origin

That's a fair point.

> unless I simply return an exception to raise here, which defeats the clarity of the [re-]raise at the end.

That's a capital suggestion!  If the helper function takes the original exception as an argument and returns an exception for the caller to raise, then, if it cannot come up with something better, the function can return the original exception -- if the caller raises that from within the `except` block which defined it, then I think that's the equivalent of a bare `raise` statement, which is exactly what we want!

Reply:
I was worried that `raise e` might not be exactly that same as `raise`. It seems there's some controversy about this point but the consensus seems to be that the original Python 2 problem (overwriting the exception origin) were fixed in Python 3 and this shouldn't be a problem.

I'm tired of playing leapfrog on this PR today, but if I don't get my approvals I guess I take a whack at this tomorrow. ;-)

Reply:
> I don't get my approvals I guess I take a whack at this tomorrow. ;-)

That creates a perverse incentive....  :-D